File: AtCoder/Greens/python/ABC012-D-Bus_to_Sakerarenai_Unmei.py
INF = float('inf')
N, M = map(int, input().split())
d = [[0 if i == j else INF for j in range(N)] for i in range(N)]
for _ in range(M):
    a, b, t = map(int, input().split())
    a -= 1
    b -= 1
    d[a][b] = t
    d[b][a] = t


def warshall_floyd():
    global d
    for k in range(N):
        for i in range(N):
            for j in range(N):
                d[i][j] = min(d[i][j], d[i][k] + d[k][j])


warshall_floyd()
res = INF
for i in range(N):
    worst = 0
    for j in range(N):
        worst = max(worst, d[i][j])
    res = min(res, worst)
print(res)

File: AtCoder/Greens/cpp/ABC012-D-Bus_to_Sakerarenai_Unmei.cpp
#include <bits/stdc++.h>
#define MAX_N 300
#define MAX_M 44850
#define INF 1'000'000'000
using namespace std;
typedef long long ll;

ll N, M;
ll d[MAX_N][MAX_N];

void warshall_floyd() {
  for (ll k = 0; k < N; ++k) {
    for (ll i = 0; i < N; ++i) {
      for (ll j = 0; j < N; ++j) {
        d[i][j] = min(d[i][j], d[i][k] + d[k][j]);
      }
    }
  }
}

int main() {
  cin >> N >> M;
  for (ll i = 0; i < N; ++i) {
    for (ll j = 0; j < N; ++j) {
      if (i == j)
        d[i][j] = 0;
      else
        d[i][j] = INF;
    }
  }
  for (ll i = 0; i < M; ++i) {
    ll a, b, t;
    cin >> a >> b >> t;
    --a;
    --b;
    d[a][b] = t;
    d[b][a] = t;
  }
  warshall_floyd();
  ll res = INF;
  for (ll i = 0; i < N; ++i) {
    ll worst = 0;
    for (ll j = 0; j < N; ++j) {
      if (i != j) worst = max(worst, d[i][j]);
    }
    res = min(res, worst);
  }
  printf("%lld\n", res);
}